db/common/src/main/java/org/commonjava/indy/db/common/AbstractStoreDataManager.java

```diff
@@ -466,19 +474,11 @@ public void disableNotValidStore(ArtifactStore store,ArtifactStoreValidateData v
         Set<StoreKey> processed = new HashSet<>();
         final String packageType = toProcess.get( 0 ).getPackageType();
 
-        Set<ArtifactStore> all = this.getAllArtifactStores().stream().filter( st -> {
-            if ( packageType != null && !st.getPackageType().equals( packageType ) )
-            {
-                return false;
-            }
-
-            if ( st.getType() != StoreType.group )
-            {
-                return false;
-            }
-
-            return true;
-        } ).collect( Collectors.toSet() );
+        Set<ArtifactStore> all = this.getStoreKeysByPkgAndType( packageType, group )
+                                     .stream()
+                                     .map( this::getArtifactStoreInternal )
```

Comment:
Doesn't it smell like what you just fixed? I suspect the getAxxx will bring performance issues. It looks to me than getting all things in a query is fast but getting the keys and then retrieve store by key is much slower.

Reply:
If you check https://github.com/Commonjava/indy/pull/1472, you will find this fix is just same to it. This fix is a add-back of some further refactor missing. 
But our current solution will not use this method actually, because IspnDataManager has override its caller with reverse map solution. I just added it back here to make the code not missing the former implementation.

Reply:
I see. It looks the getArtifactStore inside the map() does the trick. If put in for-loop it get called many many times. Do you know why it reduces so much call count?

Reply:
Actually the call count is reduced but not as much as what we see. I think the most possible reason for the call count reducing is caused by the thread-off. It is not included in the store process itself but split out to other threads.
The reverse map and the store partition has reduced the call count also. Now there is no full scan of the whole store data table, because the reverse map only stores its parent groups, and calculate based on the groups and groups' parents. I think these groups' number is much smaller than the full groups calculation, but still some there for each single calculation. (maybe tens vs thousands before)

Reply:
So it is not just #1472 but all the fixes put together to get a fast result. Perhaps the reverse map plays a key role in it. Am I right?

Reply:
#1472 is focusing on fixing the promote timeout problem which happened after 1.9.7.1 release. It can make promote performance back to the state before 1.9.7.1, but can not fix upload latency problem. 
I think two parts plays the key role of the throughout improvements: the async of the groupAffectedBy and the reversemap. The async one has completely kicked the bottleneck out of the main download/upload process, so it will not impact the throughout now. And the reverse map is focusing on reducing the call count now, and it can reduce the time of the execution itself, but it is now running in separated threads, so we don't care about it now.

Reply:
Brilliant! I think we made a correct decision last night not just release #1472. :)